## Homework #2 - Statistical Analysis. Pandas. Numpy

### Overview:
One of the main goals of classification machine learning models is to find
patterns that will allow the algorithm to separate the classes of samples.
Usually, they do that using different approaches. People can do that too,
and as a data scientist, it is essential to find manually the patterns that
will allow a Machine Learning algorithm to do that. Usually, we use for
that Statistical tools and methods.

### Tasks:
* Create a new Jupiter notebook.
* Study each column of the selected Data Set

* We highly recommend exploring the following statistical parameters:
  * Mean.
  * Median.
  * Mode.
  * Standard Deviation.
  * Variation.
  * Range of values.
* Find what columns have the highest absolute correlation with the target column (target column is specified in the Data Set page), and explain
why you think so happens. (here, you can explore the literature on the
topic).

For the start, I import Pandas Library and assign to it an alias "pd", that is conventionally used in ML and Data Science, and overall makes it easier to operate with.

In [1]:
import pandas as pd

I decided to use the "Titanic - Machine Learning from Disaster" Dataset, since it has several obstacles that will make it more interesting to work with when analyzing its contents,
such as: string data, like names, that are not relevant to the analysis itself, but makes it impossible to analyze the dataset using implemented methods and that enables me to perform
some operations in order to exclude them in order to make the methods work.

I read the CSV file using read_csv() method that will transform the CSV data in Dataframe for a easier manipulation with it.

In [2]:
dataset = pd.read_csv("dataset/titanic_train.csv")

Next, I will display the dataset using 2 methods:
* head(n) - display first n rows from the dataset
* tail(n) - display last n rows from the dataset.

In [3]:
dataset.head(n=10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [4]:
dataset.tail(10)

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

The features in the dataset are the following:
* PassengerId - useless, an ID assign to each row in the dataset,
* Survived - contains the Target Variables, that shows if the Passenger survived,
* Pclass - contains information about Ticket class,
* Name - pretty straight-forward, the Name and Surname of the Passenger,
* Sex - gender of the Passenger,
* Age - in years,
* SibSp - number of siblings / spouses of the Passenger on the Titanic,
* Parch - number of parent / children of the Passenger on the Titanic,
* Ticket - Ticket Number,
* Fare - how much money did the Passenger spent on the Ticket,
* Cabin - cabin number that Passenger stayed,
* Embarked - in which Port did Passenger embarked.

First step, I will call the method that will provide basic information about the whole dataset - info()

In [5]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


As it follows, info() returned me some information about the dataset:
* Class - Dataframe, obviously a Table Data object,
* Index Range - 891 entries (Passengers and their information), indexed 0 through 890,
* Number of Rows - 891
* Number of Columns - 12 (1 column is Target Variable - "Survived", 11 columns are Features)
* Non-Null Values - per each column, it displayed how many Non-Null values are present, for example: for PassengerId - 891 Non-Null values, therefore, no null values in that column,
or Age with 714 Non-Null values, therefore there are some Null values in Age column, and so on.
* Data type - per each column, it also showed the datatype that is stored in that column, for example: Survived contains only int64 object, or Name - contains Object, that, as we
saw, are strings, and so on. Per general, there are 2 float64 columns, 5 int64 columns and 5 object columns.
* Memory Usage - 83.7+ KB needed space to store that Data Frame object.

Next, I will check the missing values in the dataset using isnull() method from pandas, that will show a boolean for each cell (True if Null, False otherwise), and combine it with
sum() that will return a Series, that will have Indexes as the name of the columns in the original DataFrame object, and will have per each column the number of Null values.

In [6]:
print(dataset.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


There are Null values only in the column of Age, Cabin and Embarked.

Next, I will pass through each of the columns and get rid of Non-Numerical Columns or Columns that has no effect on the Survived column, that may be logically explained without
evidence, such as: PassengerId, because it is a unique identifier for each row in the dataset, Name, Ticket. About Cabin, I found that Cabin Code is the placement on the Ship, more
exactly, the Deck where the Cabin was placed, therefore most probably there is a correlation between the placement of the cabin and the survival rate of the passenger, but I might
just drop it as well, since I suppose that the PClass is related with the Cabin placement a well. I will try to just take the first letters of the Cabin, for example
Passenger with cabin C85, will be assigned to the group with C Cabin, thus I will maybe find a correlation with the actual status if Passenger survived or not.

In [7]:
dataset = dataset.drop(["PassengerId", "Name", "Ticket"], axis=1)
dataset = dataset.dropna(subset=["Cabin"])
dataset

Survived  Pclass     Sex   Age  SibSp  Parch     Fare        Cabin  \
1           1       1  female  38.0      1      0  71.2833          C85   
3           1       1  female  35.0      1      0  53.1000         C123   
6           0       1    male  54.0      0      0  51.8625          E46   
10          1       3  female   4.0      1      1  16.7000           G6   
11          1       1  female  58.0      0      0  26.5500         C103   
..        ...     ...     ...   ...    ...    ...      ...          ...   
871         1       1  female  47.0      1      1  52.5542          D35   
872         0       1    male  33.0      0      0   5.0000  B51 B53 B55   
879         1       1  female  56.0      0      1  83.1583          C50   
887         1       1  female  19.0      0      0  30.0000          B42   
889         1       1    male  26.0      0      0  30.0000         C148   

    Embarked  
1          C  
3          S  
6          S  
10         S  
11         S  
..       ...  
871        S  
872        S  
879        C  
887        S  
889        C  

[204 rows x 9 columns]

Next, I will sum up the Siblings and Parents/Children together, to make a new column, that will hold the family size altogether.

In [8]:
dataset["FamilySize"] = dataset["SibSp"] + dataset["Parch"] + 1
dataset.drop(["SibSp", "Parch"], axis=1, inplace=True)
dataset

Survived  Pclass     Sex   Age     Fare        Cabin Embarked  FamilySize
1           1       1  female  38.0  71.2833          C85        C           2
3           1       1  female  35.0  53.1000         C123        S           2
6           0       1    male  54.0  51.8625          E46        S           1
10          1       3  female   4.0  16.7000           G6        S           3
11          1       1  female  58.0  26.5500         C103        S           1
..        ...     ...     ...   ...      ...          ...      ...         ...
871         1       1  female  47.0  52.5542          D35        S           3
872         0       1    male  33.0   5.0000  B51 B53 B55        S           1
879         1       1  female  56.0  83.1583          C50        C           2
887         1       1  female  19.0  30.0000          B42        S           1
889         1       1    male  26.0  30.0000         C148        C           1

[204 rows x 8 columns]

In [9]:
copy_dataset = dataset.copy()

Next, I will do the above said, I will take only the Code of the Cabin, thus making it more Categorical Data than it is right now.

In [10]:
dataset["Cabin"] = dataset["Cabin"].apply(lambda cabin_name: cabin_name[0])
dataset

Survived  Pclass     Sex   Age     Fare Cabin Embarked  FamilySize
1           1       1  female  38.0  71.2833     C        C           2
3           1       1  female  35.0  53.1000     C        S           2
6           0       1    male  54.0  51.8625     E        S           1
10          1       3  female   4.0  16.7000     G        S           3
11          1       1  female  58.0  26.5500     C        S           1
..        ...     ...     ...   ...      ...   ...      ...         ...
871         1       1  female  47.0  52.5542     D        S           3
872         0       1    male  33.0   5.0000     B        S           1
879         1       1  female  56.0  83.1583     C        C           2
887         1       1  female  19.0  30.0000     B        S           1
889         1       1    male  26.0  30.0000     C        C           1

[204 rows x 8 columns]

Next, I will find all unique values in the dataset for each column. First, I will drop the remaining Null values from the dataset, in Age and Embarked.

In [11]:
dataset.dropna(subset=["Age", "Embarked"], inplace=True)
dataset

Survived  Pclass     Sex   Age     Fare Cabin Embarked  FamilySize
1           1       1  female  38.0  71.2833     C        C           2
3           1       1  female  35.0  53.1000     C        S           2
6           0       1    male  54.0  51.8625     E        S           1
10          1       3  female   4.0  16.7000     G        S           3
11          1       1  female  58.0  26.5500     C        S           1
..        ...     ...     ...   ...      ...   ...      ...         ...
871         1       1  female  47.0  52.5542     D        S           3
872         0       1    male  33.0   5.0000     B        S           1
879         1       1  female  56.0  83.1583     C        C           2
887         1       1  female  19.0  30.0000     B        S           1
889         1       1    male  26.0  30.0000     C        C           1

[183 rows x 8 columns]

In [12]:
original_dataset = dataset.copy()

Then, I will apply the unique() function in order to obtain all the uniques values in each column. This will enable me to correctly encode / map the values that are string to
integers, such as 0 and 1 and so on.

In [13]:
unique_values = dataset.apply(pd.Series.unique)
unique_values = unique_values.apply(lambda x : sorted(x))
print(unique_values)

Survived                                                 [0, 1]
Pclass                                                [1, 2, 3]
Sex                                              [female, male]
Age           [0.92, 1.0, 2.0, 3.0, 4.0, 6.0, 11.0, 14.0, 15...
Fare          [0.0, 5.0, 7.65, 8.05, 10.4625, 10.5, 12.475, ...
Cabin                                  [A, B, C, D, E, F, G, T]
Embarked                                              [C, Q, S]
FamilySize                                   [1, 2, 3, 4, 5, 6]
dtype: object


As it may be seen, there are several columns that may be mapped to integers. I will be using map() method in order to map each Category to a number.

In [14]:
dataset["Sex"] = dataset["Sex"].map({"female": 0, "male": 1})
dataset["Deck"] = dataset["Cabin"].map({"A": 7, "B": 6, "C": 5, "D": 4, "E": 3, "F": 2, "G": 1, "T": 0})
dataset.drop(columns=["Cabin"], axis=1, inplace=True)
dataset["Embarked"] = dataset["Embarked"].map({"C": 0, "Q": 1, "S": 2})
dataset

Survived  Pclass  Sex   Age     Fare  Embarked  FamilySize  Deck
1           1       1    0  38.0  71.2833         0           2     5
3           1       1    0  35.0  53.1000         2           2     5
6           0       1    1  54.0  51.8625         2           1     3
10          1       3    0   4.0  16.7000         2           3     1
11          1       1    0  58.0  26.5500         2           1     5
..        ...     ...  ...   ...      ...       ...         ...   ...
871         1       1    0  47.0  52.5542         2           3     4
872         0       1    1  33.0   5.0000         2           1     6
879         1       1    0  56.0  83.1583         0           2     5
887         1       1    0  19.0  30.0000         2           1     6
889         1       1    1  26.0  30.0000         0           1     5

[183 rows x 8 columns]

In [15]:
dataset.describe()

Survived      Pclass         Sex         Age        Fare    Embarked  \
count  183.000000  183.000000  183.000000  183.000000  183.000000  183.000000   
mean     0.672131    1.191257    0.519126   35.674426   78.682469    1.278689   
std      0.470725    0.515187    0.501005   15.643866   76.347843    0.957294   
min      0.000000    1.000000    0.000000    0.920000    0.000000    0.000000   
25%      0.000000    1.000000    0.000000   24.000000   29.700000    0.000000   
50%      1.000000    1.000000    1.000000   36.000000   57.000000    2.000000   
75%      1.000000    1.000000    1.000000   47.500000   90.000000    2.000000   
max      1.000000    3.000000    1.000000   80.000000  512.329200    2.000000   

       FamilySize        Deck  
count  183.000000  183.000000  
mean     1.939891    4.573770  
std      1.110239    1.476471  
min      1.000000    0.000000  
25%      1.000000    3.500000  
50%      2.000000    5.000000  
75%      2.000000    6.000000  
max      6.000000    7.000000

By the use of describe() method, I can notice that after all the manipulations with the dataset, such as dropping Rows with Null values, I got 183 rows. Also, I may conclude some of
the following:
* Only 2 possible outcomes - Survived or Not, more than half of the passengers survived the Titanic crash (around 67%),
* Only 3 possible classes of Tickets (1, 2, 3), most of the Passengers had Ticket of Class 1,
* Only 2 possible Genders - Female or Male, most people were Males, since it was encoded as Females = 0 and Males = 1, and 50% shows that the value in the sorted column is 1.0,
therefore slightly more Males than Females.
* The range of the age is between 1 year and 80 years old, the average age in the dataset is 35, median 36, showing that .
* The average fare for the trip was 78, the minimum being 0, probably because of workers, and highest 512. Standard deviation shows 76, which means that other fares are widely
spread around the mean, meaning there was a high variance of people that could afford higher fare and those who did not. Also, it shows that most people on the Titanic, were people
from middle class, since 78 fare is between the 25% and 75% quantiles.
* Most people embarked at Southampton, since the 50% quantile shows the value of 2, the range being 0 to 2 (Cherbourg, Queenstown and Southampton)
* The average family size was of 2 persons, the largest family had 6 people in it, also there were passengers that travelled alone,
* The most populated Deck was Deck 5 and 6 - D, E, taken in count that average fare was 78, most probably the lower Decks were more populated with middle ranked people. There were
Decks from A, B, C, D, E, F, G, T, T being the lower Deck with the machines and Technical staff, A being the upper Deck.

In [16]:
dataset.mode()

Survived  Pclass  Sex   Age   Fare  Embarked  FamilySize  Deck
0         1       1    1  36.0  26.55         2           1     5

As it may be noticed, I used mode() method from pandas, that shows the most frequent value in each column. As I mentioned above, most frequent value for:
* Survived column is 1
* PClass is 1
* Sex is 1, which is Male
* Age is 36, which is close to 35 mentioned above
* Fare - 26.55, very different from the mentioned above, due to the fact that there is a very high standard deviation, meaning that there may be multiple common values of Fare
* Embarked - 2, which is Southampton, as mentioned above,
* Family Size - 1, again, maybe due to the fact that Standard Deviation is 1.11, and the range between 0% and 25% shows that 1/4 of the people were alone and range between 25% and
75% shows that 3/4 travelled either alone or with another member, making it average of 2
* Deck - 5, as mentioned Deck D was most frequent among Passengers description.

In [17]:
copy_dataset.dropna(subset=["Age", "Embarked"], inplace=True)

In [18]:
copy_dataset = pd.get_dummies(copy_dataset, columns=["Sex"], dtype='int')
copy_dataset

Survived  Pclass   Age     Fare        Cabin Embarked  FamilySize  \
1           1       1  38.0  71.2833          C85        C           2   
3           1       1  35.0  53.1000         C123        S           2   
6           0       1  54.0  51.8625          E46        S           1   
10          1       3   4.0  16.7000           G6        S           3   
11          1       1  58.0  26.5500         C103        S           1   
..        ...     ...   ...      ...          ...      ...         ...   
871         1       1  47.0  52.5542          D35        S           3   
872         0       1  33.0   5.0000  B51 B53 B55        S           1   
879         1       1  56.0  83.1583          C50        C           2   
887         1       1  19.0  30.0000          B42        S           1   
889         1       1  26.0  30.0000         C148        C           1   

     Sex_female  Sex_male  
1             1         0  
3             1         0  
6             0         1  
10            1         0  
11            1         0  
..          ...       ...  
871           1         0  
872           0         1  
879           1         0  
887           1         0  
889           0         1  

[183 rows x 9 columns]

In [19]:
copy_dataset["Cabin"] = copy_dataset["Cabin"].apply(lambda x : x[0])
copy_dataset = pd.get_dummies(copy_dataset, columns=["Cabin"], prefix="Deck", dtype='int')
copy_dataset

Survived  Pclass   Age     Fare Embarked  FamilySize  Sex_female  \
1           1       1  38.0  71.2833        C           2           1   
3           1       1  35.0  53.1000        S           2           1   
6           0       1  54.0  51.8625        S           1           0   
10          1       3   4.0  16.7000        S           3           1   
11          1       1  58.0  26.5500        S           1           1   
..        ...     ...   ...      ...      ...         ...         ...   
871         1       1  47.0  52.5542        S           3           1   
872         0       1  33.0   5.0000        S           1           0   
879         1       1  56.0  83.1583        C           2           1   
887         1       1  19.0  30.0000        S           1           1   
889         1       1  26.0  30.0000        C           1           0   

     Sex_male  Deck_A  Deck_B  Deck_C  Deck_D  Deck_E  Deck_F  Deck_G  Deck_T  
1           0       0       0       1       0       0       0       0       0  
3           0       0       0       1       0       0       0       0       0  
6           1       0       0       0       0       1       0       0       0  
10          0       0       0       0       0       0       0       1       0  
11          0       0       0       1       0       0       0       0       0  
..        ...     ...     ...     ...     ...     ...     ...     ...     ...  
871         0       0       0       0       1       0       0       0       0  
872         1       0       1       0       0       0       0       0       0  
879         0       0       0       1       0       0       0       0       0  
887         0       0       1       0       0       0       0       0       0  
889         1       0       0       1       0       0       0       0       0  

[183 rows x 16 columns]

In [20]:
copy_dataset = pd.get_dummies(copy_dataset, columns=["Embarked"], dtype='int')
copy_dataset

Survived  Pclass   Age     Fare  FamilySize  Sex_female  Sex_male  \
1           1       1  38.0  71.2833           2           1         0   
3           1       1  35.0  53.1000           2           1         0   
6           0       1  54.0  51.8625           1           0         1   
10          1       3   4.0  16.7000           3           1         0   
11          1       1  58.0  26.5500           1           1         0   
..        ...     ...   ...      ...         ...         ...       ...   
871         1       1  47.0  52.5542           3           1         0   
872         0       1  33.0   5.0000           1           0         1   
879         1       1  56.0  83.1583           2           1         0   
887         1       1  19.0  30.0000           1           1         0   
889         1       1  26.0  30.0000           1           0         1   

     Deck_A  Deck_B  Deck_C  Deck_D  Deck_E  Deck_F  Deck_G  Deck_T  \
1         0       0       1       0       0       0       0       0   
3         0       0       1       0       0       0       0       0   
6         0       0       0       0       1       0       0       0   
10        0       0       0       0       0       0       1       0   
11        0       0       1       0       0       0       0       0   
..      ...     ...     ...     ...     ...     ...     ...     ...   
871       0       0       0       1       0       0       0       0   
872       0       1       0       0       0       0       0       0   
879       0       0       1       0       0       0       0       0   
887       0       1       0       0       0       0       0       0   
889       0       0       1       0       0       0       0       0   

     Embarked_C  Embarked_Q  Embarked_S  
1             1           0           0  
3             0           0           1  
6             0           0           1  
10            0           0           1  
11            0           0           1  
..          ...         ...         ...  
871           0           0           1  
872           0           0           1  
879           1           0           0  
887           0           0           1  
889           1           0           0  

[183 rows x 18 columns]

I decided to use One-Hot Encoding, because Sex, Deck and Embarked are Non-Ordinal Categorical Data, thus, it will help me make more accurate decisions when I will analyze the data.

In [21]:
copy_dataset.corr()

Survived    Pclass       Age      Fare  FamilySize  Sex_female  \
Survived    1.000000 -0.034542 -0.254085  0.134241    0.077731    0.532418   
Pclass     -0.034542  1.000000 -0.306514 -0.315235   -0.027821    0.046181   
Age        -0.254085 -0.306514  1.000000 -0.092424   -0.274985   -0.184969   
Fare        0.134241 -0.315235 -0.092424  1.000000    0.431089    0.130433   
FamilySize  0.077731 -0.027821 -0.274985  0.431089    1.000000    0.121397   
Sex_female  0.532418  0.046181 -0.184969  0.130433    0.121397    1.000000   
Sex_male   -0.532418 -0.046181  0.184969 -0.130433   -0.121397   -1.000000   
Deck_A     -0.050109 -0.098613  0.155518 -0.138541   -0.145124   -0.210777   
Deck_B      0.085066 -0.206306 -0.050394  0.278442    0.030087    0.085704   
Deck_C     -0.111087 -0.231388  0.016425  0.220260    0.176847    0.011597   
Deck_D      0.067153 -0.054695  0.097199 -0.120913   -0.080742    0.061023   
Deck_E      0.057730  0.093722  0.069318 -0.185130   -0.122603   -0.042134   
Deck_F     -0.019268  0.532319 -0.254816 -0.197038    0.034493   -0.059340   
Deck_G     -0.054813  0.526266 -0.200495 -0.127816    0.109376    0.155319   
Deck_T     -0.106131 -0.027593  0.044308 -0.042040   -0.062924   -0.071342   
Embarked_C  0.104870 -0.231837  0.085018  0.240382   -0.083460    0.062691   
Embarked_Q -0.038544 -0.039131  0.019038  0.015625    0.053177    0.004024   
Embarked_S -0.095861  0.238755 -0.088568 -0.242172    0.071434   -0.063146   

            Sex_male    Deck_A    Deck_B    Deck_C    Deck_D    Deck_E  \
Survived   -0.532418 -0.050109  0.085066 -0.111087  0.067153  0.057730   
Pclass     -0.046181 -0.098613 -0.206306 -0.231388 -0.054695  0.093722   
Age         0.184969  0.155518 -0.050394  0.016425  0.097199  0.069318   
Fare       -0.130433 -0.138541  0.278442  0.220260 -0.120913 -0.185130   
FamilySize -0.121397 -0.145124  0.030087  0.176847 -0.080742 -0.122603   
Sex_female -1.000000 -0.210777  0.085704  0.011597  0.061023 -0.042134   
Sex_male    1.000000  0.210777 -0.085704 -0.011597 -0.061023  0.042134   
Deck_A      0.210777  1.000000 -0.146812 -0.164661 -0.119633 -0.117303   
Deck_B     -0.085704 -0.146812  1.000000 -0.344483 -0.250282 -0.245406   
Deck_C     -0.011597 -0.164661 -0.344483  1.000000 -0.280710 -0.275241   
Deck_D     -0.061023 -0.119633 -0.250282 -0.280710  1.000000 -0.199974   
Deck_E      0.042134 -0.117303 -0.245406 -0.275241 -0.199974  1.000000   
Deck_F      0.059340 -0.066992 -0.140153 -0.157192 -0.114207 -0.111982   
Deck_G     -0.155319 -0.039600 -0.082846 -0.092918 -0.067509 -0.066194   
Deck_T      0.071342 -0.019636 -0.041080 -0.046075 -0.033475 -0.032823   
Embarked_C -0.062691  0.126285  0.154229  0.022545  0.060548 -0.174436   
Embarked_Q -0.004024 -0.027846 -0.058257  0.169113 -0.047472 -0.046547   
Embarked_S  0.063146 -0.119444 -0.140641 -0.058894 -0.049904  0.183333   

              Deck_F    Deck_G    Deck_T  Embarked_C  Embarked_Q  Embarked_S  
Survived   -0.019268 -0.054813 -0.106131    0.104870   -0.038544   -0.095861  
Pclass      0.532319  0.526266 -0.027593   -0.231837   -0.039131    0.238755  
Age        -0.254816 -0.200495  0.044308    0.085018    0.019038   -0.088568  
Fare       -0.197038 -0.127816 -0.042040    0.240382    0.015625   -0.242172  
FamilySize  0.034493  0.109376 -0.062924   -0.083460    0.053177    0.071434  
Sex_female -0.059340  0.155319 -0.071342    0.062691    0.004024   -0.063146  
Sex_male    0.059340 -0.155319  0.071342   -0.062691   -0.004024    0.063146  
Deck_A     -0.066992 -0.039600 -0.019636    0.126285   -0.027846   -0.119444  
Deck_B     -0.140153 -0.082846 -0.041080    0.154229   -0.058257   -0.140641  
Deck_C     -0.157192 -0.092918 -0.046075    0.022545    0.169113   -0.058894  
Deck_D     -0.114207 -0.067509 -0.033475    0.060548   -0.047472   -0.049904  
Deck_E     -0.111982 -0.066194 -0.032823   -0.174436   -0.046547    0.183333  
Deck_F      1.000000 -0.037804 -0.018745   -0.187693   -0.026583    0.192194  
Deck_G     -0.03

Analyzing the Correlation Matrix, I came to the following conclusions:
* Females survived more than males, that having the highest impact on the Survival Rate of the Passengers, probably because Males risked their lives more to save females and children
rather than saving themselves,
* The class has a very small impact on the Survival Rate, but has high impact on the Deck F, G, thus being due to the fact that lower the Class of the Ticket (in mapping 3rd
class is 3, 2nd - 2, 1st - 1), were placed at lower decks, with worse conditions, while people with Ticket class 1, were placed at deck A, B, C, and people in the middle class were
placed at deck D and E. Also, has a inverse proportionate correlation with the Fare, thus 1st class tickets are more expensive than 3rd class ones. At the same time, Older people
had better class than other people.
* Age is inverse proportionate to the survival rate, thus people of greater age were more likely to not Survive the crash, while Children were more likely to survive it, probably
because they were saved first, before the older people,
* People with at least one family member were more likely to survive the crash, probably because they had other people that could save them, or help each other
* People on Deck T had the higher mortality rate, while people on Decks D, E, B, that were of middle to upper class were more likely to survive, probably because they were at higher
levels and had access to upper deck faster, while Deck C was more crowded, since it was a General Transitional Deck, with multiple classes.
* Embarkation Port does not play a significant role in the Survival Rate, which is obvious, since it may not have an effect on the crash cause - Iceberg.

In [22]:
original_dataset.groupby(['Pclass'])["Survived"].value_counts()

Pclass  Survived
1       1           106
        0            52
2       1            12
        0             3
3       0             5
        1             5
Name: count, dtype: int64

Since the Dataset is pretty skewed (as we saw with the Median, Mode and Mean values), for example, the most people were in the 1st class, thus, it is pretty inconclusive to say that
people of lower class were more likely to not survive, but taken in count the statistics of the available data, this seems to visualize a little bit the correlation.

In [23]:
original_dataset.groupby(['Embarked'])["Survived"].value_counts()

Embarked  Survived
C         1           48
          0           17
Q         0            1
          1            1
S         1           74
          0           42
Name: count, dtype: int64

In [24]:
original_dataset.groupby(['Cabin'])["Survived"].value_counts()

Cabin  Survived
A      1            7
       0            5
B      1           32
       0           11
C      1           30
       0           21
D      1           23
       0            8
E      1           22
       0            8
F      1            7
       0            4
G      0            2
       1            2
T      0            1
Name: count, dtype: int64

The Deck played a role in the survival rate pretty similar to what I mentioned above, the people on lower decks were more like not to survive the crash, while in the middle Decks,
they were more safe.

In [25]:
correlations = dataset.corr()["Survived"].sort_values(ascending=False)
print(f"Correlations \n{correlations}\n")
correlations = dataset.corr()["Survived"].abs().sort_values(ascending=False)
print(f"Absolute Correlations \n{correlations}")

Correlations 
Survived      1.000000
Fare          0.134241
FamilySize    0.077731
Deck          0.019181
Pclass       -0.034542
Embarked     -0.100943
Age          -0.254085
Sex          -0.532418
Name: Survived, dtype: float64

Absolute Correlations 
Survived      1.000000
Sex           0.532418
Age           0.254085
Fare          0.134241
Embarked      0.100943
FamilySize    0.077731
Pclass        0.034542
Deck          0.019181
Name: Survived, dtype: float64


## Conclusions:
* People in higher classes were more likely to survive,
* Females were highly more likely to survive than males,
* Younger people were more likely to survive,
* Respectively, people that payed more Fare, were more likely to survive.